src/include/miopen/env.hpp

```diff
@@ -31,127 +31,166 @@
-            }
+            return false;
+        }
+        else if(value_env_str == "enable" || value_env_str == "enabled" || value_env_str == "1" ||
+                value_env_str == "yes" || value_env_str == "on" || value_env_str == "true")
+        {
+            return true;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusInvalidValue, "Invalid value for env variable");
         }
-        return (std::strcmp(value_env_str.c_str(), "enable") == 0 ||
-                std::strcmp(value_env_str.c_str(), "enabled") == 0 ||
-                std::strcmp(value_env_str.c_str(), "1") == 0 ||
-                std::strcmp(value_env_str.c_str(), "yes") == 0 ||
-                std::strcmp(value_env_str.c_str(), "on") == 0 ||
-                std::strcmp(value_env_str.c_str(), "true") == 0);
+
+        return false; // shouldn't reach here
     }
-}
+};
+
+// Supports hexadecimals (with leading "0x"), octals (if prefix is "0") and decimals (default).
+// Returns 0 if environment variable is in wrong format (strtoull fails to parse the string).
+template <>
+struct ParseEnvVal<uint64_t>
+{
+    static uint64_t go(const char* vp) { return std::strtoull(vp, nullptr, 0); }
+};
 
-// Return 0 if env is enabled else convert environment var to an int.
-// Supports hexadecimal with leading 0x or decimal
-inline uint64_t EnvvarValue(const char* name, uint64_t fallback = 0)
+template <>
+struct ParseEnvVal<std::string>
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
+    static std::string go(const char* vp) { return std::string{vp}; }
+};
+
+template <typename T>
+struct EnvVar
+{
+private:
+    T value{};
+    bool is_unset = true;
+
+public:
+    const T& GetValue() const { return value; }
+
+    bool IsUnset() const { return is_unset; }
+
+    void UpdateValue(const T& val)
     {
-        return fallback;
+        is_unset = false;
+        value    = val;
     }
-    else
+
+    explicit EnvVar(const char* const name, const T& def_val)
     {
-        return strtoull(value_env_p, nullptr, 0);
+        // NOLINTNEXTLINE (concurrency-mt-unsafe)
+        const char* vp = std::getenv(name);
+        if(vp != nullptr) // a value was provided
+        {
+            is_unset = false;
+            value    = ParseEnvVal<T>::go(vp);
+        }
+        else // no value provided, use default value
+        {
+            value = def_val;
+        }
     }
+};
+
+} // end namespace internal
+
+// static inside function hides the variable and provides
+// thread-safety/locking
+#define MIOPEN_DECLARE_ENV_VAR(name, type, default_val)                    \
+    struct name                                                            \
+    {                                                                      \
+        using value_type = type;                                           \
+        static miopen::internal::EnvVar<type>& Ref()                       \
```

Comment:
@JehandadKhan @cderb @amberhassaan 
As member function is defined in the class body, the function is implicitly inlined by the compiler. Which means that each translation unit (module) that uses `MIOPEN_DECLARE_ENV_VAR(A)` will contain an instance of A::Ref() together with static variable in it. 

However, if this is so, then it is impossible to change a cached environment variable from a test, because tests resides in one translation unit, while variable is used (read) in some other one.

I was somewhat concerned about this problem, and even wrote a small test to figure it out. In short, the program only uses one instance of a static member function (for ex. clang linker takes the function from the module that is encountered first). So it it _seems_ like there is no problem so far.

[Notice] There is no ODR violation provided that the class definitions are identical. Otherwise we run into the UB but cppcheck is able to detect this.

⚠️ **_However, we should be very careful._** For example, if in one module we place `MIOPEN_DECLARE_ENV_VAR` in the namespace `miopen`, and in another module we “define” an environment variable - in the global one, then we will get two classes (and each will use its own static member function).

In order to avoid this potential problem, I can recommend splitting the macro to "declaration" and "definition" (notice that default_val is not necessary in declaration anymore):
```C++
#define MIOPEN_DECLARE_ENV_VAR(name, type)             \
    struct name                                        \
    {                                                  \
        using value_type = type;                       \
        static miopen::internal::EnvVar<type>& Ref();  \
    };

#define MIOPEN_DEFINE_ENV_VAR(name, type, default_val)                 \
    miopen::internal::EnvVar<type>& name::Ref()                        \
    {                                                                  \
        static miopen::internal::EnvVar<type> var{#name, default_val}; \
        return var;                                                    \
    }
```
The programmer can put the "declaration" macro in multiple files, in headers etc, but "definition" should appear only once in the program. If some declaration resides in wrong namespace, then the build would fail as definition of the static member function won't be found. And that would guarantee that exactly one instance of the cached env variable exists in the program.

Reply:
This is a decent compromise but I'm not sure if we should do this work right now as it seems unnecessary. As you said, `function scope static` ensures that there is only one definition of the `static var`. Things may have been different for `class scope static`. 

One thing to note is that `env vars` defined in our main source tree will get combined into `libMIOpen.so` first and will get instantiated there first in most cases. 
Many of our `definitions` are in `.cpp` files but we will need some in `.hpp` files for the ones that we wish to update in our `gtests`. 

@JehandadKhan , @cderb: This last point is essential to our current work and may require moving the definitions of some variables to `header` files in order for them to be read/updated from `gtests`.

Reply:
@amberhassaan Let me repeat: If in one module we put MIOPEN_DECLARE_ENV_VAR(X) on one namespace, and in some other module in another namespace, then we'll have two separate cached values. For example, the second module won't be able to change the value cached in the first module. Silently. No errors, no warnings.

> I'm not sure if we should do this work right now as it seems unnecessary

Yeah, it is absolutely unnecessary provided that we will always use MIOPEN_DECLARE_ENV_VAR() in the same namespace. Which means another implicit dependence that each programmer from now on should bear in mind.

Reply:
I disagree. 1) Any competent programmer would notice that the `env var` is being `declared` in two places, which would not seem right
2) If this becomes a concern and a pit that many programmers are falling into, I can think of other ways to handle this situation. One such way is to cache the environment variables by name in an internal global `set` or `map`. This has the benefit of checking if the `env` variable has been declared. Secondly, this will make setting `env vars` in tests easier but slightly more expensive (which we don't care about because tests are expected to set once).

Reply:
@atamazov what if we add namespace miopen as part of the macro for MIOPEN_DECLARE_ENV_VAR? Could this concern be resolved?

Reply:
that most likely won't work because we have different namesaces nested inside miopen

Reply:
@amberhassaan of course, and C++ does not have means to "pushd" to the root namespace and then "popd".

Reply:
@atamazov @cderb @amberhassaan @JehandadKhan 
Another way is to move all the declarations into one header and make all variables visible everywhere.
So there will be no namespace problems anymore. And the unique static instance is guaranteed by C++.

Reply:
We discussed this in the daily standup and identified more benefits of this: 
1) `namespace` will be standardized for all environment variables. Duplicate definitions will be easier to detect. I suggest that we put them in `miopen::env`. 
2) We can document the variables in the comments as well for quick reference. 

Cons (that are minor) are that:
1) All `env vars` are exposed because all are in headers. This may be a good thing because `gtests` won't have to worry about whether `env var` can be accessed from `gtest code`. 
2) Uses are farther away from definitions. 

IMO, pros outweigh the cons and we should do this in a follow up PR.

Reply:
Just to be more specific and to avoid any misunderstandings or concerns.
The following code is correct (no UB or any violations) and works as expected - there is only one static variable:
`header.h`
```C++
struct A
{
    static A& ref()
    {
        static A val{};
        return val;
    }
    int i{0};
};

extern int foo(); 
```
`main.cpp`
```C++
#include "header.h"
#include <cstdio>
int main()
{
    printf("%d\n", A::ref().i);
    printf("%d\n", foo());
    printf("%d\n", ++A::ref().i);
    printf("%d\n", foo());
    printf("%d\n", A::ref().i);
}
```
`foo.cpp`
```C++
#include "header.h"
int foo()
{
    A::ref().i++;
    return A::ref().i;
}
```
`output`
```
0
1
2
3
3
```
The only problem is the namespaces, but if we move all the `MIOPEN_DECLARE_ENV_VAR` into one big header we can solve it without overcomplications.
It may affect compilation time and probably there is a point that particular variables should be placed together with the code which uses them, but anyways, there is no perfect solution anyway. 

More serious problem is "non-constness" of static variable and complex non-default initialization which leads to explicit check of guard variable on each access.

Reply:
As about a simple check of unique instance in a particular namespace. It is `static_assert(std::is_same_v<>)`

For example
```C++
#include <type_traits>

#define MIOPEN_DECLARE_ENV_VAR(name, type, default_val)                    \
    struct name                                                            \
    {                                                                      \
        static_assert(std::is_same_v<name, ::miopen::env::name>, "must be in miopen::env and must be unique");\
        static type& Ref() { static type var{default_val}; return var; }   \
    }; 

namespace miopen::env{ 
MIOPEN_DECLARE_ENV_VAR(XXX, long, 321)
}

//namespace miopen::solver::env{ 
//MIOPEN_DECLARE_ENV_VAR(XXX, long, 321)
//}
```
this won't be compiled if `MIOPEN_DECLARE_ENV_VAR` is not in `::miopen::env` namespace.

And my second concern about checking of guard variable on each access can be ignored. Current develop implementation behaves in the same way - it initializes the static variable on a first call and check initialization guard value on each access, so we shouldn't see any performance degradation.
Anyway, changing that kind of functionality and forcing early initialization may affect library loading time, so let's keep everything as it is.
Probably the essential variables like logging can be modified to be initialized earlier, but it's an another story.

Reply:
@CAHEK7 Very goo idea.

[Recommendation] We can add `namespace miopen::env{ ... }` into the macro and ask engineers to use the macro only in the global namespace. Just to reduce typing.

Reply:
[resolved]